File: backend/app/schemas/contracts.py
"""SHTAPM canonical shared data contract (D006 / D007) — P0 stub.

Authoritative field names and shapes come from **Doc05 §05.8** (see
project-state/DECISIONS.md D007). PRD §10.3 shorthand (``temp``/``vib``,
``s1..s6``, nested ``healing:{}``) is explicitly superseded.

Scope: schema / validation stub ONLY. No MQTT, no simulator, no ML, no business
logic. This module is the single Python source of truth for the wire contract;
``frontend/src/types/contracts.ts`` mirrors it verbatim, and future
edge/simulator/backend code consume these models.

Envelope (ruling E): MQTT canonical payloads carry NO ``type`` field — the MQTT
topic (telemetry/decision/ledger/status/command) identifies the category. The
WebSocket layer wraps a payload as ``{"type": <category>, **payload}``. The
payload fields are otherwise identical on MQTT and WebSocket.

Not frozen here (out of M2 scope): ``device_status`` / ``system_health`` /
``alert`` WS frames (Doc05 §05.8) — frozen in P4; and the ``command`` payload,
which is UNSPECIFIED in the docs (U14) and must not be invented.
"""

from __future__ import annotations

from enum import Enum
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field

# Frozen six-channel order (ruling A/B). Names match Doc05 sensors.channel ENUM,
# sensor_readings columns, and the WS telemetry/decision frames.
CHANNELS: tuple[str, ...] = (
    "temperature",
    "vibration",
    "pressure",
    "humidity",
    "gas",
    "current",
)


class Channel(str, Enum):
    temperature = "temperature"
    vibration = "vibration"
    pressure = "pressure"
    humidity = "humidity"
    gas = "gas"
    current = "current"


class Attribution(str, Enum):  # Doc05 decisions.attribution
    none = "none"
    fault = "fault"
    attack = "attack"


class HealthState(str, Enum):  # Doc05 decisions.health_state (wire field: "health")
    healthy = "healthy"
    warning = "warning"
    critical = "critical"


class RLAction(str, Enum):  # Doc05 decisions.rl_action
    continue_ = "continue"  # "continue" is a Python keyword; value stays canonical
    reduce_weight = "reduce_weight"
    isolate = "isolate"
    alert = "alert"
    safe_stop = "safe_stop"


# Trust and severity are scores in [0, 1] (FR-T1; Doc05 "0–1").
Score = Annotated[float, Field(ge=0.0, le=1.0)]


class _Strict(BaseModel):
    """Reject unknown/renamed fields so the frozen contract can't silently drift."""

    model_config = ConfigDict(extra="forbid")


class SensorReadings(_Strict):
    temperature: float
    vibration: float
    pressure: float
    humidity: float
    gas: float
    current: float


class TelemetryMessage(_Strict):
    device_id: str
    ts: str  # ISO-8601 with ms (FR-Q2); kept as string on the wire
    sensors: SensorReadings
    sample_seq: int


class AnomalyInfo(_Strict):
    flag: bool
    severity: Score
    attribution: Attribution
    reason: str


class TrustScores(_Strict):
    temperature: Score
    vibration: Score
    pressure: Score
    humidity: Score
    gas: Score
    current: Score


class DecisionMessage(_Strict):
    device_id: str
    ts: str
    anomaly: AnomalyInfo
    trust: TrustScores
    health: HealthState
    failure_eta: float  # cycles/seconds ahead (Doc05); short-horizon (FR-M2)
    rl_action: RLAction
    isolated: list[Channel]  # ruling C — flat arrays, not nested healing:{}
    substituted: list[Channel]


class LedgerMessage(_Strict):
    device_id: str
    ts: str
    block_index: int
    event: str  # stored as ledger_blocks.event_type; wire field name is "event"
    payload_hash: str  # ruling D — kept (required by hash-chain verification)
    prev_hash: str
    this_hash: str


class WSFrameType(str, Enum):
    telemetry = "telemetry"
    decision = "decision"
    ledger = "ledger"
    device_status = "device_status"
    system_health = "system_health"
    alert = "alert"

File: frontend/src/types/contracts.ts
// SHTAPM canonical shared data contract (D006 / D007) — mirror of
// backend/app/schemas/contracts.py. Authoritative field names/shapes: Doc05
// §05.8. Keep this file in EXACT sync with the Python source (the contract is
// sacred, TRD §02.6). PRD §10.3 shorthand (temp/vib, s1..s6, nested healing)
// is superseded. MQTT payloads carry no `type`; the WS layer wraps them.

export const CHANNELS = [
  "temperature",
  "vibration",
  "pressure",
  "humidity",
  "gas",
  "current",
] as const;
export type Channel = (typeof CHANNELS)[number];

export type Attribution = "none" | "fault" | "attack";
export type HealthState = "healthy" | "warning" | "critical";
export type RLAction = "continue" | "reduce_weight" | "isolate" | "alert" | "safe_stop";

export interface SensorReadings {
  temperature: number;
  vibration: number;
  pressure: number;
  humidity: number;
  gas: number;
  current: number;
}

export interface TelemetryMessage {
  device_id: string;
  ts: string; // ISO-8601 with ms
  sensors: SensorReadings;
  sample_seq: number;
}

export interface AnomalyInfo {
  flag: boolean;
  severity: number; // 0..1
  attribution: Attribution;
  reason: string;
}

// Same six channel keys as SensorReadings; values are trust scores in 0..1.
export type TrustScores = SensorReadings;

export interface DecisionMessage {
  device_id: string;
  ts: string;
  anomaly: AnomalyInfo;
  trust: TrustScores;
  health: HealthState;
  failure_eta: number;
  rl_action: RLAction;
  isolated: Channel[]; // ruling C — flat arrays, not nested healing:{}
  substituted: Channel[];
}

export interface LedgerMessage {
  device_id: string;
  ts: string;
  block_index: number;
  event: string;
  payload_hash: string; // ruling D — required
  prev_hash: string;
  this_hash: string;
}

// WebSocket envelope (ruling E): type is a WS concern; MQTT payloads omit it.
export type WSFrameType =
  | "telemetry"
  | "decision"
  | "ledger"
  | "device_status"
  | "system_health"
  | "alert";
export type WSFrame<T> = { type: WSFrameType } & T;

// Canonical valid examples (also give tsc a compile-time shape check).
export const EXAMPLE_TELEMETRY: TelemetryMessage = {
  device_id: "pump-01",
  ts: "2026-08-09T12:00:00.000Z",
  sensors: { temperature: 24.5, vibration: 0.03, pressure: 1013.2, humidity: 41.0, gas: 120.0, current: 0.42 },
  sample_seq: 123,
};

export const EXAMPLE_DECISION: DecisionMessage = {
  device_id: "pump-01",
  ts: "2026-08-09T12:00:03.000Z",
  anomaly: { flag: true, severity: 0.82, attribution: "attack", reason: "pressure vs current" },
  trust: { temperature: 0.95, vibration: 0.93, pressure: 0.21, humidity: 0.9, gas: 0.88, current: 0.97 },
  health: "warning",
  failure_eta: 142,
  rl_action: "isolate",
  isolated: ["pressure"],
  substituted: ["pressure"],
};

export const EXAMPLE_LEDGER: LedgerMessage = {
  device_id: "pump-01",
  ts: "2026-08-09T12:00:03.100Z",
  block_index: 57,
  event: "isolate",
  payload_hash: "9f0c",
  prev_hash: "7a3b",
  this_hash: "a1b2",
};
